components/resource_control/src/resource_group.rs

```diff
@@ -0,0 +1,402 @@
+/// duration to update the minimal priority value of each resource group.
+pub const MIN_PRIORITY_UPDATE_INTERVAL: Duration = Duration::from_secs(1);
+/// default resource group name
+const DEFAULT_RESOURCE_GROUP_NAME: &str = "default";
+
+pub enum ResourceConsumeType {
+    CpuTime(Duration),
+    IoBytes(u64),
+}
+
+/// ResourceGroupManager manages the metadata of each resource group.
+pub struct ResourceGroupManager {
+    resource_groups: DashMap<String, GroupSettings>,
+    registry: Mutex<Vec<Arc<ResourceController>>>,
+    /// total_ru_quota is an estimated upper bound of RU, used to calculate the
+    /// weight of each resource.
+    total_ru_quota: f64,
+}
+
+impl Default for ResourceGroupManager {
+    fn default() -> Self {
+        let total_ru_quota = SysQuota::cpu_cores_quota() * 10000.0;
+        Self {
+            resource_groups: DashMap::new(),
+            registry: Mutex::new(vec![]),
+            total_ru_quota,
+        }
+    }
+}
+
+impl ResourceGroupManager {
+    fn get_ru_setting(setting: &GroupSettings, is_read: bool) -> f64 {
+        if setting.get_mode() == GroupMode::RuMode {
+            if is_read {
+                setting.get_r_u_settings().get_r_r_u().get_tokens()
+            } else {
+                setting.get_r_u_settings().get_w_r_u().get_tokens()
+            }
+        } else if is_read {
+            setting.get_resource_settings().get_cpu().get_tokens()
+        } else {
+            setting.get_resource_settings().get_io_write().get_tokens()
+        }
+    }
+
+    fn gen_group_priority_factor(&self, setting: &GroupSettings, is_read: bool) -> u64 {
+        let ru_settings = Self::get_ru_setting(setting, is_read);
+        // TODO: ensure the result is a valid positive integer
+        (self.total_ru_quota / ru_settings * 10.0) as u64
+    }
+
+    pub fn add_resource_group(&self, config: ResourceGroup) {
+        let group_name = config.get_name().to_ascii_lowercase();
+        self.registry.lock().unwrap().iter().for_each(|controller| {
+            let priority_factor =
+                self.gen_group_priority_factor(config.get_settings(), controller.is_read);
+            controller.add_resource_group(group_name.clone().into_bytes(), priority_factor);
+        });
+        self.resource_groups
+            .insert(group_name, config.get_settings().clone());
+    }
+
+    pub fn remove_resource_group(&self, name: &str) {
+        let group_name = name.to_ascii_lowercase();
+        // do not remove the default resource group.
+        if DEFAULT_RESOURCE_GROUP_NAME == group_name {
+            return;
+        }
+        self.registry.lock().unwrap().iter().for_each(|controller| {
+            controller.remove_resource_group(group_name.as_bytes());
+        });
+        self.resource_groups.remove(&group_name);
+    }
+
+    pub fn get_resource_group(&self, name: &str) -> Option<Ref<'_, String, GroupSettings>> {
+        self.resource_groups.get(&name.to_ascii_lowercase())
```

Comment:
how about restricting or converting on tidb side, it's wasteful to convert it to lowercase everytime

Reply:
tidb will convert into lowercase then put it to resource manager(PD). That's mean the group name is case-insensitive.

Reply:
This method should be only use to update or get the resource group meta, so it is not called very often. All methods under `ResourceController` won't do this convert because they are under the hot path.